src/TestableIO.System.IO.Abstractions.TestingHelpers/Events/MockFileSystemEvents.cs

```diff
@@ -0,0 +1,394 @@
+        {
+            try
+            {
+                sub.Handler(args);
+            }
+            catch (Exception ex)
+            {
+                exceptions.Add(ex);
+            }
+        }
+        
+        if (exceptions.Count > 0)
+        {
+            if (exceptions.Count == 1)
+            {
+                throw exceptions[0];
+            }
+            throw new AggregateException(
+                $"One or more event handlers failed for {operation} operation on {path}", 
+                exceptions);
+        }
+        
+        // Handle responses only for Before phase
+        if (phase != OperationPhase.Before)
+        {
+            return;
+        }
+        var response = args.GetResponse();
+        if (response == null)
+        {
+            return;
+        }
+
+        if (response.Exception != null)
+        {
+            throw response.Exception;
+        }
+
+        if (response.Cancel)
+        {
+            throw new OperationCanceledException(
+                $"Operation {operation} on {path} was cancelled.");
+        }
+    }
+
+    /// <summary>
+    /// Removes the specified subscription from the list of active subscriptions.
+    /// </summary>
+    /// <param name="subscription">The subscription to be removed.</param>
+    private void RemoveSubscription(Subscription subscription)
+    {
+        lock (LockObject)
+        {
+            if (!subscriptions.Remove(subscription))
+            {
+                return;
+            }
+            Interlocked.Increment(ref subscriptionVersion);
+        }
+    }
+    
+#if FEATURE_SERIALIZABLE
+    [Serializable]
+#endif
+    private class Subscription : IDisposable
+    {
+#if FEATURE_SERIALIZABLE
+        /// <summary>
+        /// Represents a reference to the parent <see cref="MockFileSystemEvents"/> instance
+        /// that manages the subscriptions and events for the file system operations.
+        /// </summary>
+        [NonSerialized]
+#endif
+        private readonly MockFileSystemEvents parent;
+        /// <summary>
+        /// Specifies the set of file system operations that this subscription filters on.
+        /// </summary>
+        /// <remarks>
+        /// When set, only the specified file operations will trigger the event handler.
+        /// If null, all file system operations are considered for this subscription.
+        /// </remarks>
+        private readonly HashSet<FileOperation> filterOperations;
+
+        /// <summary>
+        /// Indicates whether the subscription has been disposed.
+        /// </summary>
+        private volatile bool isDisposed;
+        
+#if FEATURE_SERIALIZABLE
+        [NonSerialized]
+#endif
+        public readonly Action<FileSystemOperationEventArgs> Handler;
+        /// <summary>
+        /// Gets a value indicating whether this subscription has been disposed.
+        /// </summary>
+        public bool IsDisposed
```

Comment:
As you remove the `Subscription` from the collection, it seems unnecessary to provide a flag and check for it.
What are your reasons for doing so?